informalsystems/hermes
# ICS29: Add `fee transfer` command

Closes: #2714

See https://github.com/informalsystems/ibc-rs/pull/1981#discussion_r979819760

## Description

Add a `fee transfer` command which can be used to exercise the relayer incentivization support introduced in #1981.

### Testing

In order to test the `fee transfer` command, open 3 terminals:

__First terminal__

Setup the test chains

1. Run `nix shell github:informalsystems/cosmos.nix#ibc-go-v5-simapp` or `nix shell --extra-experimental-features nix-command --extra-experimental-features flakes github:informalsystems/cosmos.nix#ibc-go-v5-simapp`
2. `cd` to the `hermes` repo
3. Start the test chains with `CHAIN_COMMAND_PATH=simd cargo run --bin test_setup_with_fee_enabled_binary_channel`

__Second terminal__

Start a Hermes instance

1. Run `nix shell github:informalsystems/cosmos.nix#ibc-go-v5-simapp` or `nix shell --extra-experimental-features nix-command --extra-experimental-features flakes github:informalsystems/cosmos.nix#ibc-go-v5-simapp`
2. `cd` to the `hermes` repo
3. Run `source data/test/binary-channels.env`
4. Update `max_msg_num = 30` to `max_msg_num = 4` in the file `relayer-config.toml` located in `data/test/`
5. Run `CHAIN_COMMAND_PATH=simd cargo run --bin hermes -- --config $RELAYER_CONFIG start`

__Third terminal__

Transfer some tokens with fees. In addition this tests that if the pay message is not in the same batch as the transfer message, the relayer still receives the fees

1. `cd` to the `hermes` repo
2. Run `source data/test/binary-channels.env`
3. Check the balance of the relayers and users, in this case the only ones which will change are the relayer on chain A, user1 on chain A and user2 on chain B:
  - `gaiad --node $NODE_A_RPC_ADDR query bank balances $NODE_A_WALLETS_RELAYER_ADDRESS`
  - `gaiad --node $NODE_A_RPC_ADDR query bank balances $NODE_A_WALLETS_USER1_ADDRESS`
  - `gaiad --node $NODE_B_RPC_ADDR query bank balances $NODE_B_WALLETS_USER2_ADDRESS`
4. Transfer 30 times 5321 `samoleans` from `user1` to `user2`:
  -  `cargo run --bin hermes -- --config $RELAYER_CONFIG fee transfer --dst-chain $CHAIN_ID_B --src-chain $CHAIN_ID_A --src-port $PORT_A --src-channel $CHANNEL_ID_A --amount 5321 --denom $NODE_A_DENOM --receive-fee 1233 --ack-fee 772 --timeout-fee 2001 --timeout-seconds 120 --recipient $NODE_B_WALLETS_USER2_ADDRESS --key-name user1 --number-msgs 30`
5. Check the balances:
  - The relayer on chain A should have received 30 times 2005 `samoleans` (60150)
  - `user1` on chain A should have lost 30 times 2005+5321 `samoleans` (219780)
  - `user2` on chain B should have received 30 times 5321 `ibc/<hash>` (159630) 

______

### PR author checklist:
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:
